distributed/shuffle/_worker_extension.py

```diff
@@ -996,16 +1000,24 @@ def split_by_partition(t: pa.Table, column: str) -> dict[Any, pa.Table]:
     return dict(zip(partitions, shards))
 
 
-def convert_chunk(data: bytes, subdims: tuple[int, ...]) -> np.ndarray:
+def convert_chunk(data: bytes) -> np.ndarray:
     import numpy as np
 
     from dask.array.core import concatenate3
 
     file = BytesIO(data)
-    rec_cat_arg = np.empty(subdims, dtype="O")
+    shards: dict[NIndex, np.ndarray] = {}
+
     while file.tell() < len(data):
-        subindex, subarray = pickle.load(file)
-        rec_cat_arg[tuple(subindex)] = subarray
+        index, shard = pickle.load(file)
+        shards[index] = shard
+
+    subshape = [max(dim) + 1 for dim in zip(*shards.keys())]
+    assert len(shards) == np.prod(subshape)
+
+    rec_cat_arg = np.empty(subshape, dtype="O")
+    for index, shard in shards.items():
+        rec_cat_arg[tuple(index)] = shard
     del data
     del file
     arrs = rec_cat_arg.tolist()
```

Comment:
TODO: it feels naively like it is very wasteful to make objects of everything, convert to a nested list and then concatenate. Why is it not possible to just allocate the right size of thing up front and insert directly?